node/network/dispute-distribution/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+/// `DisputeSender` manages the sending side of all initiated disputes.
+///
+/// See the implementers guide for more detail on the general protocol, but in general
+/// `DisputeSender` takes care of getting our vote out to all other relevant validators.
+mod sender;
+use self::sender::{DisputeSender, FromSendingTask};
+
+/// Handle receival of dispute requests.
+///
+/// - Spam/Flood handling
+/// - Trigger import of statements
+mod receiver;
+use self::receiver::DisputesReceiver;
+
+/// Error and [`Result`] type for this subsystem.
+mod error;
+use error::{Fatal, FatalResult};
+use error::{Result, log_error};
+
+// mod metrics;
+//// Prometheus `Metrics` for dispute distribution.
+// pub use metrics::Metrics;
+
+// #[cfg(test)]
+// mod tests;
+
+const LOG_TARGET: &'static str = "parachain::dispute-distribution";
+
+/// The dispute distribution subsystem.
+pub struct DisputeDistributionSubsystem {
+	/// Easy and efficient runtime access for this subsystem.
+	runtime: RuntimeInfo,
+
+	/// Sender for our dispute requests.
+	disputes_sender: DisputeSender,
+
+	/// Receive messages from `SendTask`.
+	sender_rx: mpsc::Receiver<FromSendingTask>,
+}
+
+impl<Context> Subsystem<Context> for DisputeDistributionSubsystem
+where
+	Context: SubsystemContext<Message = DisputeDistributionMessage> + Sync + Send,
+{
+	fn start(self, ctx: Context) -> SpawnedSubsystem {
+		let future = self
+			.run(ctx)
+			.map_err(|e| SubsystemError::with_origin("dispute-distribution", e))
+			.boxed();
+
+		SpawnedSubsystem {
+			name: "dispute-distribution-subsystem",
+			future,
+		}
+	}
+}
+
+impl DisputeDistributionSubsystem {
+
+	/// Create a new instance of the availability distribution.
+	pub fn new(keystore: SyncCryptoStorePtr) -> Self {
+		let runtime = RuntimeInfo::new_with_config(runtime::Config {
+			keystore: Some(keystore),
+			session_cache_lru_size: DISPUTE_WINDOW as usize,
+		});
+		let (tx, sender_rx) = mpsc::channel(1);
+		let disputes_sender = DisputeSender::new(tx);
+		Self { runtime, disputes_sender, sender_rx }
+	}
+
+	/// Start processing work as passed on from the Overseer.
+	async fn run<Context>(mut self, mut ctx: Context) -> std::result::Result<(), Fatal>
+	where
+		Context: SubsystemContext<Message = DisputeDistributionMessage> + Sync + Send,
+	{
+
+		loop {
+			let message = Message::receive(&mut ctx, &mut self.sender_rx).await;
+			match message {
+				Message::Subsystem(result) => {
+					let result = match result? {
+						FromOverseer::Signal(signal) => {
+							match self.handle_signals(&mut ctx, signal).await {
+								SignalResult::Conclude => return Ok(()),
+								SignalResult::Result(result) => result,
+							}
+						}
+						FromOverseer::Communication { msg } =>
+							self.handle_subsystem_message(&mut ctx, msg).await,
+					};
+					log_error(result, "on FromOverseer")?;
+				}
+				Message::Sender(result) => {
+					self.disputes_sender.on_task_message(
+						result.ok_or(Fatal::SenderExhausted)?
+					).await;
+				}
+			}
```

Comment:
nit: `fn _inner` to flatten the pyramid a bit

Reply:
The pyramid looks way worse on github than in the editor. Longest line length is 79, it is also very little code - another indirection would not really improve things, I believe.